File: src/commands/cycleCommand.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open CommandUtils

let print_endlinef = Utils_js.print_endlinef

let prerr_endlinef = Utils_js.prerr_endlinef

let spec =
  {
    CommandSpec.name = "cycle";
    doc = "Output .dot file for cycle containing the given file";
    usage =
      Printf.sprintf
        "Usage: %s cycle [OPTION]...\n\ne.g. %s cycle path/to/file.js \n"
        Utils_js.exe_name
        Utils_js.exe_name;
    args =
      CommandSpec.ArgSpec.(
        empty
        |> base_flags
        |> connect_flags
        |> root_flag
        |> strip_root_flag
        |> flag "--types" no_arg ~doc:"Only consider type dependencies"
        |> anon "FILE..." (required string));
  }

let main base_flags option_values root strip_root types_only file () =
  let flowconfig_name = base_flags.Base_flags.flowconfig_name in
  let file = expand_path file in
  let root = guess_root flowconfig_name root in
  let strip_root f =
    if strip_root then
      Files.relative_path (Path.to_string root) f
    else
      f
  in
  (* connect to server *)
  let request = ServerProt.Request.CYCLE { filename = file; types_only } in
  match connect_and_make_request flowconfig_name option_values root request with
  | ServerProt.Response.CYCLE (Error msg) -> FlowExitStatus.(exit ~msg Unknown_error)
  | ServerProt.Response.CYCLE (Ok dep_graph) ->
    (* print .dot file to stdout *)
    LwtUtils.output_graph Lwt_io.stdout strip_root dep_graph |> Lwt_main.run
  | response -> failwith_bad_response ~request ~response

let command = CommandSpec.command spec main

File: rust_port/crates/flow_cli/src/cycle_command.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use flow_command_spec::arg_spec;
use flow_server_env::server_prot;

pub(crate) fn spec() -> flow_command_spec::Spec {
    let spec = flow_command_spec::Spec::new(
        "cycle",
        "Output .dot file for cycle containing the given file",
        flow_command_spec::Visibility::Public,
        format!(
            "Usage: {} cycle [OPTION]...\n\ne.g. {} cycle path/to/file.js \n",
            flow_command_utils::exe_name(),
            flow_command_utils::exe_name()
        ),
    );
    let spec = flow_command_utils::add_base_flags(spec);
    let spec = flow_command_utils::add_connect_flags(spec);
    let spec = flow_command_utils::add_root_flag(spec);
    let spec = flow_command_utils::add_strip_root_flag(spec);
    spec.flag(
        "--types",
        &arg_spec::truthy(),
        "Only consider type dependencies",
        None,
    )
    .anon("FILE...", &arg_spec::required(None, arg_spec::string()))
}

pub(crate) fn run(args: &arg_spec::Values) {
    let base_flags = flow_command_utils::get_base_flags(args);
    let connect_flags = flow_command_utils::get_connect_flags(args);
    let root_arg =
        flow_command_spec::get(args, "--root", &arg_spec::optional(arg_spec::string())).unwrap();
    let strip_root = flow_command_spec::get(args, "--strip-root", &arg_spec::truthy()).unwrap();
    let types_only = flow_command_spec::get(args, "--types", &arg_spec::truthy()).unwrap();
    let file = flow_command_spec::get(
        args,
        "FILE...",
        &arg_spec::required(None, arg_spec::string()),
    )
    .unwrap();

    let file = flow_command_utils::expand_path(&file);
    let root = flow_command_utils::guess_root(&base_flags.flowconfig_name, root_arg.as_deref());
    let strip_root = |f: &str| {
        if strip_root {
            flow_common::files::relative_path(&root, f)
        } else {
            f.to_string()
        }
    };
    // connect to server
    let request = server_prot::request::Command::CYCLE {
        filename: file,
        types_only,
    };
    let response = flow_command_utils::connect_and_make_request(
        &base_flags.flowconfig_name,
        &connect_flags,
        &root,
        &request,
    );
    match response {
        server_prot::response::Response::CYCLE(Err(msg)) => {
            eprintln!("{}", msg);
            flow_common_exit_status::exit(flow_common_exit_status::FlowExitStatus::UnknownError);
        }
        server_prot::response::Response::CYCLE(Ok(dep_graph)) => {
            // print .dot file to stdout
            println!("digraph {{");
            for (from, dep_fs) in dep_graph {
                for dep_f in dep_fs {
                    println!("  \"{}\" -> \"{}\"", strip_root(&from), strip_root(&dep_f));
                }
            }
            print!("}}");
        }
        response => {
            flow_command_utils::failwith_bad_response(&request, &response);
        }
    }
}

pub(crate) fn command() -> flow_command_spec::Command {
    flow_command_spec::command(spec(), run)
}
